# Разработать модель для предсказания <span style='color:red'>риска</span> сердечного приступа

# <span style='color:RoyalBlue'>Описание проекта</span>

<span style='font-size:18px'>
В этом проекте предстоит разработать модель для предсказания риска сердечного приступа, используя данные пациентов, такие как возраст, привычки, состояние здоровья и результаты анализа крови.
В начале проекта будет доступен обучающий датасет и тестовый датасет, не содержащий правильных ответов. Метрика на тестовом датасете будет измерена в ходе проверки работы.
Также вам будет доступен скрипт, с помощью которого будет измеряться финальная метрика вашей модели.

# <span style='color:RoyalBlue'>Цели иследования</span>

- <span style='font-size:18px'> Разработать модель для предсказания риска сердечного приступа </span>

# <span style='color:RoyalBlue'>Описание данных</span>

- <span style="font-size: 18px;">**Фаил для тренировок** - `heart_train.csv`;</span>
    - <span style='color:blue'>**Unnamed: 0**</span> —  Безымянный;      
    - <span style='color:blue'>**Age**</span> — Возраст;
    - <span style='color:blue'>**Cholesterol**</span> — Показатель холестерина;       
    - <span style='color:blue'>**Heart rate**</span> — Частота сердцебиения;   
    - <span style='color:blue'>**Diabetes**</span> — Диабет;
    - <span style='color:blue'>**Family History**</span> — Семейная история;
    - <span style='color:blue'>**Smoking**</span> — Курение;
    - <span style='color:blue'>**Obesity**</span> — Ожирение;
    - <span style='color:blue'>**Alcohol Consumption**</span> — Употребление алкоголя;
    - <span style='color:blue'>**Exercise Hours Per Week**</span> — Количество Часов упражнений в Неделю;
    - <span style='color:blue'>**Diet**</span> — Диета;
    - <span style='color:blue'>**Previous Heart Problems**</span> — Предыдущие проблемы с сердцем;
    - <span style='color:blue'>**Medication Use**</span> — Использование лекарств;
    - <span style='color:blue'>**Stress Level**</span> — Уровень стресса;
    - <span style='color:blue'>**Sedentary Hours Per Day**</span> — Часы Сидячего Образа Жизни В День;
    - <span style='color:blue'>**Income**</span> — Доход;
    - <span style='color:blue'>**BMI**</span> — Индекс массы тела;
    - <span style='color:blue'>**Triglycerides**</span> — Триглицериды;
    - <span style='color:blue'>**Physical Activity Days Per Week**</span> — Дни физической активности в Неделю;
    - <span style='color:blue'>**Sleep Hours Per Day**</span> — Количество Часов сна в Сутки;
    - <span style='color:green'>**Heart Attack Risk**</span> — Риск сердечного приступа (бинарный);
    - <span style='color:blue'>**Blood sugar**</span> — Уровень сахара в крови;
    - <span style='color:blue'>**CK-MB**</span> — CK-MB;
    - <span style='color:blue'>**Troponin**</span> — Тропонин;
    - <span style='color:blue'>**Gender**</span> — Пол;     
    - <span style='color:blue'>**Systolic blood pressure**</span> — Систолическое кровяное давление;
    - <span style='color:blue'>**Diastolic blood pressure**</span> — Диастолическое кровяное давление;
    - <span style='color:red'>**id**</span> — уникальный идентификатор клиента.

- <span style="font-size: 18px;">**Фаил для тестирование** - `heart_test.csv`.</span>
    - <span style='color:blue'>**Unnamed: 0**</span> —  Безымянный;      
    - <span style='color:blue'>**Age**</span> — Возраст;
    - <span style='color:blue'>**Cholesterol**</span> — Показатель холестерина;       
    - <span style='color:blue'>**Heart rate**</span> — Частота сердцебиения;   
    - <span style='color:blue'>**Diabetes**</span> — Диабет;
    - <span style='color:blue'>**Family History**</span> — Семейная история;
    - <span style='color:blue'>**Smoking**</span> — Курение;
    - <span style='color:blue'>**Obesity**</span> — Ожирение;
    - <span style='color:blue'>**Alcohol Consumption**</span> — Употребление алкоголя;
    - <span style='color:blue'>**Exercise Hours Per Week**</span> — Количество Часов упражнений в Неделю;
    - <span style='color:blue'>**Diet**</span> — Диета;
    - <span style='color:blue'>**Previous Heart Problems**</span> — Предыдущие проблемы с сердцем;
    - <span style='color:blue'>**Medication Use**</span> — Использование лекарств;
    - <span style='color:blue'>**Stress Level**</span> — Уровень стресса;
    - <span style='color:blue'>**Sedentary Hours Per Day**</span> — Часы Сидячего Образа Жизни В День;
    - <span style='color:blue'>**Income**</span> — Доход;
    - <span style='color:blue'>**BMI**</span> — Индекс массы тела;
    - <span style='color:blue'>**Triglycerides**</span> — Триглицериды;
    - <span style='color:blue'>**Physical Activity Days Per Week**</span> — Дни физической активности в Неделю;
    - <span style='color:blue'>**Sleep Hours Per Day**</span> — Количество Часов сна в Сутки;
    - <span style='color:blue'>**Blood sugar**</span> — Уровень сахара в крови;
    - <span style='color:blue'>**CK-MB**</span> — CK-MB;
    - <span style='color:blue'>**Troponin**</span> — Тропонин;
    - <span style='color:blue'>**Gender**</span> — Пол;     
    - <span style='color:blue'>**Systolic blood pressure**</span> — Систолическое кровяное давление;
    - <span style='color:blue'>**Diastolic blood pressure**</span> — Диастолическое кровяное давление;
    - <span style='color:red'>**id**</span> — уникальный идентификатор клиента.

# <span style='color:RoyalBlue'>Содержание</span>

- <span style='font-size: 18px;color: RoyalBlue'>Описание проекта</span>
- <span style='font-size: 18px;color: RoyalBlue'>Цели иследования</span>
- <span style='font-size: 18px;color: RoyalBlue'>Описание данных</span>
- <span style='font-size: 18px;color: RoyalBlue'>Содержание</span>
- <span style='font-size: 18px;color: Blue'>1. Подгруска необходимых модулей и настройка необходимых констант</span>
- <span style='font-size: 18px;color: Blue'>2. Загрузка данных</span>
- <span style='font-size: 18px;color: Blue'>3. Предобработка данных</span>
- <span style='font-size: 18px;color: Blue'>4. Исследовательский анализ данных</span>
- <span style='font-size: 18px;color: Blue'>5. Создание пайплайнов</span>
- <span style='font-size: 18px;color: Blue'>6. ??? </span>
- <span style='font-size: 18px;color: RoyalBlue'>Итоговые выводы</span>

## <span style='color:Blue'>Шаг. Подгруска необходимых модулей и настройка необходимых констант</span>

In [1]:
# Тихая установка нужных библиотек

#!pip install matplotlib==3.9.4 -q
#!pip install pandas==2.2.3 -q
#!pip install seaborn==0.13.2 -q
#!pip install numpy==1.23.0 -q
#!pip install numba==0.60.0 -q
#!pip install llvmlite==0.43.0 -q
#!pip install shap==0.44.0 -q
#!pip install phik==0.12.4 -q
#!pip install pyarrow==10.0.1 -q
#!pip install scikit-learn==1.6.1 -q
#!pip install sklearn-compat==0.1.3 -q
#!pip install imbalanced-learn==0.12.4 -q

#!pip freeze

In [2]:
# импортируйте библиотеку pandas
import pandas as pd
# импорт библиотеки seaborn
import seaborn as sns
# импорт библиотеки phik
#import phik
# импортируйте plt
import matplotlib.pyplot as plt
# импортируйте np
import numpy as np
# импортируйте stats
import scipy.stats as st
# импортируйте math для расчёта факториала и корня
#from math import sqrt
# Подгруска распределение датафрейма
from sklearn.model_selection import (
    train_test_split,
    RandomizedSearchCV,
    #cross_val_score,
    #GridSearchCV,
    #KFold,
)
# Подгружаем библиотеку для работы с пайплан 
from sklearn.pipeline import Pipeline
# Клас помогает работать с данными разного типа в одном наборе
from sklearn.compose import ColumnTransformer
# Подгруска для маштабирования и кодирования признаков
from sklearn.preprocessing import (
    StandardScaler,
    #OneHotEncoder,
    #OrdinalEncoder,
    #LabelEncoder,
    MinMaxScaler,
    RobustScaler,
    PolynomialFeatures,
)
# класс для работы с пропусками
#from sklearn.impute import SimpleImputer
# Подгружаем нужные модели
from sklearn.tree import  DecisionTreeRegressor, DecisionTreeClassifier
from sklearn.linear_model import LinearRegression, LogisticRegression
#from sklearn.neighbors import KNeighborsClassifier
#from sklearn.svm import SVC
# Подгружаем метрики
from sklearn.metrics import (
    #r2_score,
    #mean_squared_error,
    #mean_absolute_error,
    root_mean_squared_error,
    confusion_matrix,
    #roc_auc_score,
    #make_scorer,
)
# Для проверки адекватности лучшей модели
#from sklearn.dummy import DummyClassifier
# для работы с несбалансированными данными
from imblearn.over_sampling import RandomOverSampler
# Подгружаем библиотеку shap для оцените важность признаков
import shap
# Настройки numpy
np.seterr(divide = 'ignore');
# import warnings filter
from warnings import simplefilter
# ignore all future warnings
simplefilter(action='ignore', category=FutureWarning)

In [3]:
# Настройки pandas
pd.options.mode.chained_assignment = None
pd.options.display.float_format = '{:,.2f}'.format

In [4]:
# Создадим константы 
RANDOM_STATE = 100
TEST_SIZE = 0.25
STATE = np.random.RandomState(12345)

In [5]:
# Общий путь к папккам
folder_location = './datasets/'

## <span style='color:Blue'> Шаг. Загрузка данных</span>

**Задача 2.1 Создание финкции вывода первичной информации.**

In [6]:
def primary_information_output(dataset):
    # Функция вывод первичной информации
    dataset.info()
    print('Количество пропущеных значений:')
    display(dataset.isna().sum())
    text = 'Количество полных дублированых строк в дата фрейме'
    print(f'{text} = {dataset.duplicated().sum()}')
    display(dataset.head())

**Задача 2.2 Считайте данные из `heart_train.csv` в датафрейм и сохраните в переменную `train`.**

In [7]:
# Прочитываю csv-файл
train = pd.read_csv(f'{folder_location}heart_train.csv', sep=',')

**Задача 2.3 Выведите основную информацию о датафрейме `train`.**

In [8]:
# Показываю информацию о дата сете
primary_information_output(train)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8685 entries, 0 to 8684
Data columns (total 28 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Unnamed: 0                       8685 non-null   int64  
 1   Age                              8685 non-null   float64
 2   Cholesterol                      8685 non-null   float64
 3   Heart rate                       8685 non-null   float64
 4   Diabetes                         8442 non-null   float64
 5   Family History                   8442 non-null   float64
 6   Smoking                          8442 non-null   float64
 7   Obesity                          8442 non-null   float64
 8   Alcohol Consumption              8442 non-null   float64
 9   Exercise Hours Per Week          8685 non-null   float64
 10  Diet                             8685 non-null   int64  
 11  Previous Heart Problems          8442 non-null   float64
 12  Medication Use      

Unnamed: 0                           0
Age                                  0
Cholesterol                          0
Heart rate                           0
Diabetes                           243
Family History                     243
Smoking                            243
Obesity                            243
Alcohol Consumption                243
Exercise Hours Per Week              0
Diet                                 0
Previous Heart Problems            243
Medication Use                     243
Stress Level                       243
Sedentary Hours Per Day              0
Income                               0
BMI                                  0
Triglycerides                        0
Physical Activity Days Per Week    243
Sleep Hours Per Day                  0
Heart Attack Risk (Binary)           0
Blood sugar                          0
CK-MB                                0
Troponin                             0
Gender                               0
Systolic blood pressure  

Количество полных дублированых строк в дата фрейме = 0


Unnamed: 0  Age  Cholesterol  Heart rate  Diabetes  Family History  \
0           0 0.36         0.73        0.07      1.00            1.00   
1           1 0.20         0.33        0.05      1.00            1.00   
2           2 0.61         0.86        0.06      1.00            0.00   
3           3 0.73         0.01        0.05      0.00            0.00   
4           4 0.78         0.76        0.02      0.00            0.00   

   Smoking  Obesity  Alcohol Consumption  Exercise Hours Per Week  ...  \
0     1.00     1.00                 1.00                     0.54  ...   
1     0.00     0.00                 1.00                     0.07  ...   
2     1.00     1.00                 1.00                     0.94  ...   
3     1.00     0.00                 1.00                     0.70  ...   
4     1.00     0.00                 1.00                     0.41  ...   

   Physical Activity Days Per Week  Sleep Hours Per Day  \
0                             3.00                 0.33   
1                             3.00                 0.83   
2                             2.00                 1.00   
3                             0.00                 0.33   
4                             5.00                 1.00   

   Heart Attack Risk (Binary)  Blood sugar  CK-MB  Troponin  Gender  \
0                        0.00         0.23   0.05      0.04    Male   
1                        0.00         0.15   0.02      0.00  Female   
2                        0.00         0.23   0.05      0.04  Female   
3                        1.00         0.23   0.05      0.04  Female   
4                        1.00         0.23   0.05      0.04    Male   

   Systolic blood pressure  Diastolic blood pressure    id  
0                     0.21                      0.71  2664  
1                     0.41                      0.57  9287  
2                     0.24                      0.22  5379  
3                     0.35                      0.27  8222  
4                     0.62                      0.44  4047  

[5 rows x 28 columns]

**<span style='color:green'>Мини вывод:</span>**

- Не соответствие типов
- Пропущеные значения
- Минимизировать поля и сделать их корекние
- Хорошобы проверить дубли после чистки пропусков
- Стоит уданить поле `Unnamed: 0`

**Задача 2.4 Считайте данные из `heart_test.csv` в датафрейм и сохраните в переменную `test`.**

In [9]:
# Прочитываю csv-файл
test = pd.read_csv(f'{folder_location}heart_test.csv', sep=',')

**Задача 2.5 Выведите основную информацию о датафрейме `test`.**

In [10]:
# Показываю информацию о дата сете
primary_information_output(test)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 966 entries, 0 to 965
Data columns (total 27 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Unnamed: 0                       966 non-null    int64  
 1   Age                              966 non-null    float64
 2   Cholesterol                      966 non-null    float64
 3   Heart rate                       966 non-null    float64
 4   Diabetes                         935 non-null    float64
 5   Family History                   935 non-null    float64
 6   Smoking                          935 non-null    float64
 7   Obesity                          935 non-null    float64
 8   Alcohol Consumption              935 non-null    float64
 9   Exercise Hours Per Week          966 non-null    float64
 10  Diet                             966 non-null    int64  
 11  Previous Heart Problems          935 non-null    float64
 12  Medication Use        

Unnamed: 0                          0
Age                                 0
Cholesterol                         0
Heart rate                          0
Diabetes                           31
Family History                     31
Smoking                            31
Obesity                            31
Alcohol Consumption                31
Exercise Hours Per Week             0
Diet                                0
Previous Heart Problems            31
Medication Use                     31
Stress Level                       31
Sedentary Hours Per Day             0
Income                              0
BMI                                 0
Triglycerides                       0
Physical Activity Days Per Week    31
Sleep Hours Per Day                 0
Blood sugar                         0
CK-MB                               0
Troponin                            0
Gender                              0
Systolic blood pressure             0
Diastolic blood pressure            0
id          

Количество полных дублированых строк в дата фрейме = 0


Unnamed: 0  Age  Cholesterol  Heart rate  Diabetes  Family History  \
0           0 0.49         0.26        0.06      0.00            1.00   
1           1 0.22         0.95        0.08      1.00            0.00   
2           2 0.63         0.09        0.06      0.00            1.00   
3           3 0.46         0.57        0.06      1.00            1.00   
4           4 0.72         0.49        0.02      1.00            0.00   

   Smoking  Obesity  Alcohol Consumption  Exercise Hours Per Week  ...  \
0     1.00     1.00                 1.00                     0.36  ...   
1     0.00     1.00                 0.00                     1.00  ...   
2     1.00     1.00                 0.00                     1.00  ...   
3     1.00     1.00                 1.00                     0.44  ...   
4     1.00     0.00                 1.00                     0.51  ...   

   Triglycerides  Physical Activity Days Per Week  Sleep Hours Per Day  \
0           0.31                             1.00                 0.33   
1           0.09                             0.00                 0.17   
2           0.21                             7.00                 1.00   
3           0.16                             0.00                 0.67   
4           0.58                             5.00                 0.00   

   Blood sugar  CK-MB  Troponin  Gender  Systolic blood pressure  \
0         0.23   0.05      0.04    Male                     0.28   
1         0.23   0.05      0.04  Female                     0.70   
2         0.10   0.00      0.09    Male                     0.46   
3         0.20   0.06      0.27  Female                     0.74   
4         0.23   0.05      0.04    Male                     0.41   

   Diastolic blood pressure    id  
0                      0.37  7746  
1                      0.44  4202  
2                      0.78  6632  
3                      0.26  4639  
4                      0.40  4825  

[5 rows x 27 columns]

**<span style='color:green'>Мини вывод:</span>**

- Не соответствие типов
- Пропущеные значения
- Минимизировать поля и сделать их корекние
- Хорошобы проверить дубли после чистки пропусков
- Стоит уданить поле `Unnamed: 0`

## <span style='color:Blue'> Шаг. Предобработка данных</span>

**Задача 3.1. Предобработка данных датафрейма <span style='color:red'>train</span>**

**Задача 3.1.1 Получение информации о датафрейме**

In [11]:
# Первичная информация о датаврейме
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8685 entries, 0 to 8684
Data columns (total 28 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Unnamed: 0                       8685 non-null   int64  
 1   Age                              8685 non-null   float64
 2   Cholesterol                      8685 non-null   float64
 3   Heart rate                       8685 non-null   float64
 4   Diabetes                         8442 non-null   float64
 5   Family History                   8442 non-null   float64
 6   Smoking                          8442 non-null   float64
 7   Obesity                          8442 non-null   float64
 8   Alcohol Consumption              8442 non-null   float64
 9   Exercise Hours Per Week          8685 non-null   float64
 10  Diet                             8685 non-null   int64  
 11  Previous Heart Problems          8442 non-null   float64
 12  Medication Use      

In [12]:
# Создаём список наименование презатационых полей
list_field_train = {
    'unnamed': 'Безымянный',
    'age': 'Возраст',
    'cholesterol': 'Показатель холестерина',
    'hr': 'Частота сердцебиения',
    'diabetes': 'Диабет',
    'fh': 'Семейная история',
    'smoking': 'Курение',
    'obesity': 'Ожирение',
    'ac': 'Употребление алкоголя',
    'ehpw': 'Количество Часов упражнений в Неделю',
    'diet': 'Диета',
    'php': 'Предыдущие проблемы с сердцем',
    'mu': 'Использование лекарств',
    'sl': 'Уровень стресса',
    'sedentary_hpd': 'Часы Сидячего Образа Жизни В День',
    'Income': 'Доход',
    'bmi': 'Индекс массы тела',
    'triglycerides': 'Триглицериды',
    'padpw': 'Дни физической активности в Неделю',
    'shpd': 'Количество Часов сна в Сутки',
    'har': 'Риск сердечного приступа',
    'blood_sugar': 'Уровень сахара в крови',
    'ck-mb': 'CK-MB',
    'troponin': 'Тропонин',
    'gender': 'Пол',
    'sbp': 'Систолическое кровяное давление',
    'dbp': 'Диастолическое кровяное давление',
    'id': 'уникальный идентификатор клиента'      
}

**Задача 3.1.2 Обработка полей**

In [13]:
# Нормализую название полей
train.columns = list_field_train.keys()

In [14]:
# Удаление пропусков
train = train.dropna(subset=['diabetes'])

In [15]:
# Удаление поля
train = train.drop('unnamed', axis=1)

In [16]:
# Сдвигаем показатель на 2 знака чтобы вывести потом возраст
train['age'] = train['age'] * 100

In [17]:
# Меняем тип данных у полей
train = train.astype({'age': 'int'})
train = train.astype({'diabetes': 'bool'})
train = train.astype({'smoking': 'bool'})
train = train.astype({'obesity': 'bool'})
train = train.astype({'ac': 'bool'})
train = train.astype({'php': 'bool'})
train = train.astype({'mu': 'bool'})
train = train.astype({'har': 'bool'})

**Задача 3.1.3 Проверка количественные данные**

In [18]:
# Получаю список количественых полей
num_col_names = train.select_dtypes(['int', 'float']).columns.tolist()

In [19]:
# Вывод основныхз показателей количественных данных
train[num_col_names].describe()

age  cholesterol       hr       fh     ehpw     diet       sl  \
count 8,442.00     8,442.00 8,442.00 8,442.00 8,442.00 8,442.00 8,442.00   
mean     44.39         0.50     0.05     0.49     0.50     1.00     5.49   
std      23.33         0.29     0.02     0.50     0.29     0.81     2.87   
min       4.00         0.00     0.02     0.00     0.00     0.00     1.00   
25%      24.00         0.26     0.03     0.00     0.25     0.00     3.00   
50%      46.00         0.50     0.05     0.00     0.51     1.00     5.00   
75%      64.00         0.75     0.07     1.00     0.75     2.00     8.00   
max      85.00         1.00     0.08     1.00     1.00     2.00    10.00   

       sedentary_hpd   Income      bmi  triglycerides    padpw     shpd  \
count       8,442.00 8,442.00 8,442.00       8,442.00 8,442.00 8,442.00   
mean            0.50     0.50     0.50           0.51     3.51     0.50   
std             0.29     0.29     0.29           0.29     2.28     0.33   
min             0.00     0.00     0.00           0.00     0.00     0.00   
25%             0.25     0.24     0.25           0.26     2.00     0.17   
50%             0.49     0.50     0.49           0.51     3.00     0.50   
75%             0.75     0.75     0.74           0.76     6.00     0.83   
max             1.00     1.00     1.00           1.00     7.00     1.00   

       blood_sugar    ck-mb  troponin      sbp      dbp       id  
count     8,442.00 8,442.00  8,442.00 8,442.00 8,442.00 8,442.00  
mean          0.23     0.05      0.04     0.45     0.50 4,696.39  
std           0.07     0.07      0.06     0.17     0.17 2,704.77  
min           0.00     0.00      0.00     0.16     0.21     1.00  
25%           0.23     0.05      0.04     0.30     0.35 2,359.25  
50%           0.23     0.05      0.04     0.45     0.50 4,694.50  
75%           0.23     0.05      0.04     0.60     0.65 7,030.75  
max           1.00     1.00      1.00     0.74     0.79 9,376.00

**Задача 3.1.4 Проверка категориальных данных**

In [20]:
def list_unique_values_categorical_features(dataset: object) -> str:
    # Выводит список уникальных значений категориальных признаков
    cat_col_names = dataset.select_dtypes(['object', 'bool']).columns.tolist()
    for col in cat_col_names:
        print(f"Категория: [{col}]:")
        print(f"Уникальные значения категории:")
        print(dataset[col].unique())
        print(f"Количество значений по уникальности в категории:")
        print(dataset[col].value_counts())
        print('----------------------------------------------------<NEXT>-')

In [21]:
# Вывод уникальных значений категориальных признаков
list_unique_values_categorical_features(train)

Категория: [diabetes]:
Уникальные значения категории:
[ True False]
Количество значений по уникальности в категории:
diabetes
True     5505
False    2937
Name: count, dtype: int64
----------------------------------------------------<NEXT>-
Категория: [smoking]:
Уникальные значения категории:
[ True False]
Количество значений по уникальности в категории:
smoking
True     7622
False     820
Name: count, dtype: int64
----------------------------------------------------<NEXT>-
Категория: [obesity]:
Уникальные значения категории:
[ True False]
Количество значений по уникальности в категории:
obesity
False    4235
True     4207
Name: count, dtype: int64
----------------------------------------------------<NEXT>-
Категория: [ac]:
Уникальные значения категории:
[ True False]
Количество значений по уникальности в категории:
ac
True     5082
False    3360
Name: count, dtype: int64
----------------------------------------------------<NEXT>-
Категория: [php]:
Уникальные значения категории:
[ True 

**Задача 3.1.5 Проверка полученого результата**

In [22]:
# Поверка результата удаление пропусков
primary_information_output(train)

<class 'pandas.core.frame.DataFrame'>
Index: 8442 entries, 0 to 8684
Data columns (total 27 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   age            8442 non-null   int64  
 1   cholesterol    8442 non-null   float64
 2   hr             8442 non-null   float64
 3   diabetes       8442 non-null   bool   
 4   fh             8442 non-null   float64
 5   smoking        8442 non-null   bool   
 6   obesity        8442 non-null   bool   
 7   ac             8442 non-null   bool   
 8   ehpw           8442 non-null   float64
 9   diet           8442 non-null   int64  
 10  php            8442 non-null   bool   
 11  mu             8442 non-null   bool   
 12  sl             8442 non-null   float64
 13  sedentary_hpd  8442 non-null   float64
 14  Income         8442 non-null   float64
 15  bmi            8442 non-null   float64
 16  triglycerides  8442 non-null   float64
 17  padpw          8442 non-null   float64
 18  shpd         

age              0
cholesterol      0
hr               0
diabetes         0
fh               0
smoking          0
obesity          0
ac               0
ehpw             0
diet             0
php              0
mu               0
sl               0
sedentary_hpd    0
Income           0
bmi              0
triglycerides    0
padpw            0
shpd             0
har              0
blood_sugar      0
ck-mb            0
troponin         0
gender           0
sbp              0
dbp              0
id               0
dtype: int64

Количество полных дублированых строк в дата фрейме = 0


age  cholesterol   hr  diabetes   fh  smoking  obesity    ac  ehpw  diet  \
0   35         0.73 0.07      True 1.00     True     True  True  0.54     1   
1   20         0.33 0.05      True 1.00    False    False  True  0.07     2   
2   60         0.86 0.06      True 0.00     True     True  True  0.94     2   
3   73         0.01 0.05     False 0.00     True    False  True  0.70     0   
4   77         0.76 0.02     False 0.00     True    False  True  0.41     1   

   ...  padpw  shpd    har  blood_sugar  ck-mb  troponin  gender  sbp  dbp  \
0  ...   3.00  0.33  False         0.23   0.05      0.04    Male 0.21 0.71   
1  ...   3.00  0.83  False         0.15   0.02      0.00  Female 0.41 0.57   
2  ...   2.00  1.00  False         0.23   0.05      0.04  Female 0.24 0.22   
3  ...   0.00  0.33   True         0.23   0.05      0.04  Female 0.35 0.27   
4  ...   5.00  1.00   True         0.23   0.05      0.04    Male 0.62 0.44   

     id  
0  2664  
1  9287  
2  5379  
3  8222  
4  4047  

[5 rows x 27 columns]

**<span style='color:green'>Мини вывод:</span>**

- Есть дисбаланс в полях по количеству значений
- Было удаленно 243 записи с пустымми значениями (Оказалась они и тежи записи отсутстуют значения в нескольких полямх)

**Задача 3.2 Предобработка данных датафрейма <span style='color:red'>test</span>**